contracts/liquidator/LiquidatorV2.sol

```diff
@@ -0,0 +1,235 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import "./vendor/@uniswap/v3-core/contracts/interfaces/callback/IUniswapV3FlashCallback.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/base/PeripheryPayments.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/base/PeripheryImmutableState.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/libraries/PoolAddress.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/libraries/CallbackValidation.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/libraries/TransferHelper.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/interfaces/ISwapRouter.sol";
+
+import "../CometInterface.sol";
+import "../ERC20.sol";
+
+/**
+ * @title Liquidator contract compatible with 1inch DEX aggregator
+ * @author Compound
+ */
+contract LiquidatorV2 is IUniswapV3FlashCallback, PeripheryImmutableState, PeripheryPayments {
+    /** Errors */
+    error InsufficientBalance(uint256 available, uint256 required);
+    error Unauthorized();
+
+    /** Events **/
+    event Absorb(address indexed initiator, address[] accounts);
+    event AbsorbWithoutBuyingCollateral();
+    event Pay(address indexed token, address indexed payer, address indexed recipient, uint256 value);
+    event BuyAndSwap(
+        address indexed tokenIn,
+        address indexed tokenOut,
+        uint256 baseAmountPaid,
+        uint256 assetBalance,
+        uint256 amountOut
+    );
+
+    /// @notice Compound Comet protocol
+    CometInterface public immutable comet;
+
+    /// @notice Address to send liquidation proceeds to
+    address public immutable recipient;
+
+    /// @notice The scale for asset price calculations
+    uint256 public constant QUOTE_PRICE_SCALE = 1e18;
+
+    constructor(
+        CometInterface _comet,
+        address _factory,
+        address _WETH9,
+        address _recipient
+    ) PeripheryImmutableState(_factory, _WETH9) {
+        comet = _comet;
+        recipient = _recipient;
+    }
+
+    function balanceOfAsset(address asset) internal returns (uint256, uint256) {
+        uint256 collateralBalance = comet.getCollateralReserves(asset);
+
+        uint256 baseScale = comet.baseScale();
+
+        uint256 quotePrice = comet.quoteCollateral(asset, QUOTE_PRICE_SCALE * baseScale);
+        uint256 collateralBalanceInBase = baseScale * QUOTE_PRICE_SCALE * collateralBalance / quotePrice;
```

Comment:
I'll need to spend more time looking at this, but what's the issue with not buying the exact amount of collateral if it's strictly <= the protocol's collateral reserves? The txn should still succeed and leave dust amounts in Comet.

Reply:
it's fine to leave a little dust in the protocol. the issue is for the size of the swap that you request from 1inch.

if you request a swap for the entirety of the collateral (say 120e18 WBTC), you'll end up with a transfer error since you will have bought an amount that is every so slightly less than that amount (e.g. 119.999999999e18 WTBC).

basically, you need to make sure that you're not requesting a swap through 1inch for an amount that you don't actually have.